[REVIEW] Improved Array Conversion with CumlArrayDescriptor and Decorators [skip-ci] (#3040)
rapidsai/cuml: wiki/python/ESTIMATOR_GUIDE.md

# cuML Python Estimators Developer Guide

> This guide is meant to help developers follow the correct patterns when creating/modifying any class that inherits from `cuml.common.base.Base`.

Author: Done. Resolving comment

> 

Reviewer: It would be nice if we can start with a high-level guide: ```

A cuML Estimator must:
 * Inherit from cuml.common.Base
 * Follow the scikit-learn estimator guidelines (link)
 * Declare each array-type attribute they will compute as a class variable so they can be automatically converted to appropriate types
 * Add input type and return type annotations to public API functions OR wrap those functions explicitly with conversion decorators (see ...)
```

I think as the guide gets longer, it can be easy to get lost in it.

Author: I actually ended up reorganizing it more than intended. I liked your suggestion of moving the more advanced details to the end for both the decorators and descriptors. Allows dev's wanting a quick intro less of a read without removing any information.

Would be great if you could glance over the new organization and let me know what you think.

Start by reading the following:

1. [SkLearn Estimator Docs](https://scikit-learn.org/stable/developers/develop.html)
   1. `cuml` Estimator design follows Sklearn very closely. We will only cover portions where our design differs from this document
   2. Pay close attention to these sections as these topics have caused pain points in the past:
      1. [Instantiation](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      2. [Estimated Attributes](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      3. [`get_params` and `set_params`](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      4. [cloning](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
2. [SkLearn Docstring Guide](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation)
   1. We follow the same guidlines for speficying array-like objects, array shapes, dtypes, and default values

## Background

Some background is necessary to understand the design of estimators and how to work around any non-standard situations.

### Input and Output Types in cuML

> In `cuml` we support both ingesting and generating a variety of different n-dimensional array types. Estimators should be able to accept and return any array type. The types that are supported as of release 0.17:

Author: I was going to ask about this. Wasn't sure what the correct capitalization was. Thanks for clearing that up.

> In `cuml` we support both ingesting and generating a variety of different n-dimensional array types. Estimators should be able to accept and return any array type. The types that are supported as of release 0.17:

Author: Updated in the guide. Resolving comment

 - cuDF DataFrame or Series
 - Pandas DataFrame or Series
 - NumPy Arrays
 - Numba Device Arrays
 - CuPy arrays
 - CumlArray type (Internal to the `cuml` API only.)

When converting between types, it's important to minimize the CPU<->GPU type conversions as much as possible. Conversions such as NumPy -> CuPy or Numba -> Pandas DataFrame will incur a performance penalty as memory is copied from device to host or viceversa.

Converting between types of the same device, i.e. CPU<->CPU or GPU<->, do not have as significant of a penalty.

Finally, conversions between Numba<->CuPy<->CumlArray incur the least amount of overhead since only the device pointer is moved from one class to another.

Internally, all array's should be converted to `CumlArray` as much as possible since it is compatible with all output types and can be easily converted.

### Speficying the Array Output Type

Users can choose which array type should be returned by cuml by either: 
1. Individually setting the output_type property on an estimator class (i.e `Base(output_type="numpy")`)

> 2. Globally setting the `cuml.global_output_type`

Author: I left off the `using_output_type` context manager because all it is doing is setting and restoring `cuml.global_output_type`. I just didnt think it was fundamentally a different way of controlling the output type.

I have no problems adding it in here

> 2. Globally setting the `cuml.global_output_type`

Author: Added in `using_output_type`. Resolving

> Note: Setting the global output type will take precedence over any value in `Base.output_type`

Author: I hear what you are saying. I went back and forth on this a few times. Ultimately I went with a design that allows for maximum flexibility while respecting existing functionality, without making the design (and therefore the code) too complex. I found that using `Base.output_type` to override the `cuml.global_output_type` was pretty restrictive. For example, assuming `Base.output_type` overrides `cuml.global_output_type`:

- You can't temporarily change multiple estimators or one-off algorithms at the same time.
   - If you wanted to change the output_type temporarily (for example, to send to an sklearn algorithm) you would have to update the `output_type` of all estimators, call some function, then revert the `output_type` setting.
- `cuml.global_output_type` becomes the "default" output type, not the "global" setting.
   - This is because estimators would only care about `cuml.global_output_type` in the constructor, and only then if `output_type` wasnt specified as an argument.
   - This design might make more sense if estimator's default value could store a reference to `cuml.global_output_type` instead of copying the value during construction.
- The utility of `cuml.using_output_type()` is significantly decreased.
   - Changing the `cuml.global_output_type` temporarily with a context manager would only be useful during estimator construction
- We would need different methods to change the `output_type` internally vs. externally.
   - Currently, `cuml.using_output_type()` can be used in cython to cast a CumlArrayDescriptor if a developer needs a specific type of array. If `cuml.using_output_type()` could no longer accomplish this task, we would need a different method that works similar to the current design.

Ultimately, I prefer having `cuml.global_output_type` override `Base.output_type`. It allows you to do things like:
```python
estA = MyEstimator().fit(some_cupy_data)
estB = MyOtherEstimator().fit(some_cupy_data)

cuml.using_output_type("numpy"):
   sk_r2_score = sklearn.r2_score(estA.my_cupy_array_, estB.my_cupy_array_)

```

### Arguments to `__init__`

All estimators should match the arguments (including the default value) in `Base.__init__` and pass these values to `super().__init__()`. As of 0.17, all estimators should accept `handle`, `verbose` and `output_type`.

In addition, is recommended to force keyword arguments to prevent breaking changes if arguments are added or removed in future versions. For example, all arguments below after `*` must be passed by keyword:

```python
def __init__(self, *, eps=0.5, min_samples=5, max_mbytes_per_batch=None,
             calc_core_sample_indices=True, handle=None, verbose=False, output_type=None):
```

Finally, do not alter any input arguments to allow cloning of the estimator. See Sklearn's [section](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes) on instantiation for more info.

For example, the following `__init__` shows what **NOT** to do:
```python
def __init__(self, my_option="option1"):
   if (my_option == "option1"):
      self.my_option = 1
   else:
      self.my_option = 2
```

> This will break cloning since the value of `self.my_option` is not a valid input to `__init__`.

Author: Done. Resolving comment

### Estimator Array-Like Attributes

> Any array attribute stored in an estimator needs to be convertable to the user's desired output type. To make it easier to store arrays in a class that derives from `Base`, the `cuml.common.array_descriptor.CumlArrayDescriptor` was created.

Author: Done. Resolving comment

The `CumlArrayDescriptor` behaves different when accessed internally (from within on of `cuml`'s functions) vs. externally (as a user). Internally, it behaves exactly like a normal attribute, and will return the previous value set. Externally, the array will get converted to the users desired output type.

#### CumlArrayDescriptor Internal Functionality

To use the `CumlArrayDescriptor` in an estimator, the any array-like attributes need to be speficied by creating a `CumlArrayDescriptor` as a class variable. Other than that, developers can treat the attribute as they normally would.

Consider the following example estimator:
```python
import cupy as cp
import cuml
from cuml.common.array_descriptor import CumlArrayDescriptor

class TestEstimator(cuml.Base):

   # Class variables outside of any function
   my_cuml_array_ = CumlArrayDescriptor()
   my_cupy_array_ = CumlArrayDescriptor()
   my_other_array_ = CumlArrayDescriptor()

   def __init__(self, ...):

      # Initialize to None

>       self.my_cuml_array_ = None

Author: No, its completely up to the estimator (some do and some don't). Initializing to None will prevent an `AttributeError` from being raised if you accessed the attribute and will return True for `hasattr(my_est, "my_cuml_array_")`. I showed both setting one attribute to None and leaving one uninitialized for this reason.

      # Init with a cupy array
      self.my_cupy_array_ = cp.zeros((10, 10))

   def fit(self, X):

>       self._set_output_type(X)

Author: Done. Resolving comment

```python
my_est = TestEstimator()
my_est.fit(cp.ones((10,)))

# Access the CumlArrayDescriptorMeta value directly. No array conversion will occur
print(my_est.__dict__["my_cuml_array_"])
# Output: CumlArrayDescriptorMeta(input_type='cupy', values={'cuml': <cuml.common.array.CumlArray object at 0x7fd39174ae20>, 'numpy': array([ 0,  1,  1,  2,  2, -1, -1, ...

# Values from CumlArrayDescriptorMeta can be specifically read
print(my_est.__dict__["my_cuml_array_"].input_type)
# Output: "cupy"

# The input value can be accessed
print(my_est.__dict__["my_cuml_array_"].get_input_value())
# Output: CumlArray ...
```

#### Internal Array-Like Conversion

Internally, developers can use the same `global_output_type` functionality to convert arrays as well. This has the same benefits of lazy conversion and caching as when descriptors are used externally. For example, we could add a `score()` function to `TestEstimator`:

```python
def score(self):

   # Set the global output type to numpy
   with cuml.using_output_type("numpy"):
      # Accessing my_cuml_array_ will return a numpy array and
      # the result can be returned directly
      return np.sum(self.my_cuml_array_, axis=0)
```

> ### Estimator Functions

Author: Done. Resolving comment

> In order for estimator functions to accept a wide variety of array types and correctly return the user's desired type, a fair amount of boiler plate code is involved. These common steps were often repetitive, inefficient, and incorrectly implemented.

Author: Done. Resolving comment

> To assit with this, a set of decorators have been created to wrap estimator functions (and all `cuml` API functions as well) and perform the boiler plate actions automatically.

Reviewer: I feel like this suggests we need decorators everywhere, but then after a while we recommend using automatic conversion. Maybe address that directly here, like "cuML provides two options to address this - (1) for many common patterns used in functions like fit, predict, transform, cuml.Base can automatically perform the data conversions as long as a method has type annotations, but (2) decorators are also provided to handle more advanced cases.

Author: Excellent suggestion. I struggled with organizing some of these pieces and I like this better. Will reorganize.

Author: See this comment: https://github.com/rapidsai/cuml/pull/3040#discussion_r518970546

#### Common Functionality

All of these decorators perform the same basic steps with a few small differences. The common steps performed by each decorator is:

> 1. Set `cuml.global_output_type = "mirror"`

Author: Will do. Mainly this section is for developers who might be confused about why a certain decorator is behaving the way it is. Not needed for common usage.

> 1. Set `cuml.global_output_type = "mirror"`

Author: Moved this to the appendix

2. Set CuPy allocator to use RMM
   1. This replaces the existing decorator `@with_cupy_rmm`
   2. Unlike before, the CuPy allocator is only set once per API call
3. Set the estimator input attributes. Can be broken down into 3 steps:
   1. Set `_input_type` attribute
   2. Set `target_dtype` attribute
   3. Set `n_features` attribute
4. Call the desired function
5. Get the estimator output type. Can be broken down into 2 steps:
   1. Get `output_type`
   2. Get `output_dtype`
6. Convert the return value
   1. This will ultimately call `CumlArray.to_output(output_type=output_type, output_dtype=output_dtype)

While the above list of steps may seem excessive for every call, most functions follow this general form, but may skip a few steps depending on a couple of factors. For example, Step #3 is necessary on functions that modify the estimator's estimated attributes, such as `fit()`, but is not necessary for functions like `predict()` or `transform()`. And Step #5/6 are only necessary when returning array-like objects and are ommitted when returning any other type.

> Functionally, you can think of these decorators equivalent to the following pseudocode:

Reviewer: Yeah, I would definitely put this after showing how decorators are used and call it details or internals

Author: Will do

Author: Done. Resolving comment

   return self
```

**After: `fit()`**

```python
def fit(self, X) -> "KMeans:
   self.coef_ = input_to_cuml_array(X, order="K").array

   return self
```

## Do's And Don'ts

### **Do:** Add Typing Information to Estimator Functions

Adding the return type to estimator functions will allow the `Base` meta-class to automatically decorate functions based on their return type.

**Do this:**
```python
def fit(self, X, y, convert_dtype=True) -> "KNeighborsRegressor":
def predict(self, X, convert_dtype=True) -> CumlArray:
def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity') -> SparseCumlArray:
def predict(self, start=0, end=None, level=None) -> typing.Union[CumlArray, float]:
```

**Not this:**
```python
def fit(self, X, y, convert_dtype=True):
def predict(self, X, convert_dtype=True) :
def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity'):
def predict(self, start=0, end=None, level=None):
```

### **Do:** Return Array-Like Objects Directly

There is no need to convert the array type before returning it. Simply return any array-like object and it will be automatically converted

**Do this:**
```python
def predict(self, X, y) -> CumlArray:

>    np_arr = np.ones((10,))

Reviewer: Maybe demo with cp.ones instead so it doesn't look like we're encouraging numpy usage?

Author: Done. Resolving comment

   return np_arr
```

**Not this:**
```python
def predict(self, X, y) -> CumlArray:
   np_arr = np.ones((10,))

   cuml_arr = input_to_cuml_array(X, order="K").array

   return cuml_arr.to_output(self._get_output_type(X))
```

### Example Decorator Use Cases

The following section shows some before and after examples of how the decorators should be used with default options.

> **Example 1: 

Reviewer: Doesn't seem to format correctly, maybe *Example 1:*?

Author: Fixed. Resolving comment

Start by reading the following:

1. [SkLearn Estimator Docs](https://scikit-learn.org/stable/developers/develop.html)
   1. cuML Estimator design follows Scikit-learn very closely. We will only cover portions where our design differs from this document
   2. Pay close attention to these sections as these topics have caused pain points in the past:
      1. [Instantiation](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      2. [Estimated Attributes](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      3. [`get_params` and `set_params`](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      4. [cloning](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
2. [Scikit-learn's Docstring Guide](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation)
   1. We follow the same guidelines for specifying array-like objects, array shapes, dtypes, and default values

## Quick Start Guide

At a high level, all cuML Estimators must:

> 1. Inherit from `cuml.common.base.Base`

Reviewer: Quick start guide is definitely helpful... Thank you!

Author: Great. Glad to hear.

The full set of return types rules that will be applied by the `Base` metaclass are:

| Return Type | Converts Array Type? | Common Methods | Notes | 
| :---------: | :-----------: | :----------- | :----------- |
| `Base` | No | `fit()` | Any type that inherits or `isinstance` of `Base` will work |
| `CumlArray` | Yes | `predict()`, `transform()` | Functions can return any array-like object (`np.ndarray`, `cp.ndarray`, etc. all accepted) |
| `SparseCumlArray` | Yes | `predict()`, `transform()` | Functions can return any sparse array-like object (`scipy`, `cupyx.scipy` sparse arrays accepted) |
| `dict`, `tuple`, `list` or `typing.Union` | Yes |  | Functions must return a generic object that contains an array-like object. No sparse arrays are supported |

Simply setting the return type of a method is all that is necessary to automatically convert the return type (with the added benefit of adding more information to the code). Below are some examples to show simple methods using automatic array conversion.

##### `fit()`

```python

> def fit(self, X) -> "KMeans:

Reviewer: ```suggestion
def fit(self, X) -> "KMeans":
```

Author: Fixed

   return cp_arr
```

**Not this:**
```python
def predict(self, X, y) -> CumlArray:
   cp_arr = cp.ones((10,))

   # Don't be tempted to use `CumlArray(cp_arr)` here either
   cuml_arr = input_to_cuml_array(cp_arr, order="K").array

   return cuml_arr.to_output(self._get_output_type(X))
```

### **Don't:** Use `CumlArray.to_output()` directly

Using `CumlArray.to_output()` is no longer necessary except in very rare circumstances. Converting array types is best handled with `input_to_cuml_array` or `cuml.using_output_type()` when retrieving `CumlArrayDescriptor` values.

**Do this:**
```python
def _private_func(self) -> CumlArray:
   return cp.ones((10,))

def predict(self, X, y) -> CumlArray:

   self.my_cupy_attribute_ = cp.zeros((10,))

   with cuml.using_output_type("numpy"):
      np_arr = self._private_func()

      return self.my_cupy_attribute_ + np_arr
```

**Not this:**
```python
def _private_func(self) -> CumlArray:
   return cp.ones((10,))

def predict(self, X, y) -> CumlArray:

   self.my_cupy_attribute_ = cp.zeros((10,))

   np_arr = CumlArray(self._private_func()).to_output("numpy")

   return CumlArray(self.my_cupy_attribute_).to_output("numpy") + np_arr
```

> ### **Don't:** Perform parameter validation in `__init__()`

Reviewer: I would call this "parameter modification" or "transformation" -- validation (checking) should be fine.

Author: Done

| `set_output_dtype` | `bool` | `False` | Whether to call `_set_target_dtype(target_arg)` |
| `set_n_features_in` | `bool` | Varies | Whether to call `_set_n_features_in(input_arg)` |
| `get_output_type` | `bool` | Varies | Whether to call `_get_output_type(input_arg)` |
| `get_output_dtype` | `bool` | `False` | Whether to call `_get_target_dtype()` |

An example of how these arguments can be used is below:

**Before:**
```python
@with_cupy_rmm
def predict(self, X, y):
   # Determine the output type and dtype
   out_type = self._get_output_type(y)
   out_dtype = self._get_target_dtype()

   # Convert to CumlArray
   X_m = input_to_cuml_array(X, order="K").array

   # Call a cuda function
   someCudaFunction(X_m.ptr)

   # Convert the CudaArray to the desired output
   return X_m.to_output(output_type=out_type, output_dtype=out_dtype)
```

**After:**
```python
@cuml.internals.api_base_return_array(input_arg="y", get_output_dtype=True)
def predict(self, X):
   # Convert to CumlArray
   X_m = input_to_cuml_array(X, order="K").array

   # Call a cuda function
   someCudaFunction(X_m.ptr)

   # Convert the CudaArray to the desired output
   return X_m
```

> #### Before `0.17` and After Comparison

Reviewer 2: Love how you moved the section, thanks! It makes is super clear for existing devs to see pre-0.17/post0.17 behavior, love it.

Author: Great. thanks.

# cuML Python Estimators Developer Guide

This guide is meant to help developers follow the correct patterns when creating/modifying any cuML Estimator object and ensure a uniform cuML API.

**Note:** This guide is long, because it includes internal details on how cuML manages input and output types for advanced use cases. But for the vast majority of estimators, the requirements are very simple and can follow the example patterns shown below in the [Quick Start Guide](#quick-start-guide).

> Start by reading the following:

Reviewer 2: If this is required as part of the quick start guide then I would move it there, otherwise I would move it to just after the quick start guide. Otherwise there's already a _lot_ of information here presented before the quick start guide which might cost a reader that doesn't need it significant time.

Author: Would be good to discuss this offline. I think we feel differently about how much information is before the quick start guide. Moving it after the guide seems awkward to me.

Reviewer: I get Dante's point and have generally supported getting the overview near the start, but I don't think this is a problem to leave in as it is important background.

Author: Talked with Dante. Going to change it from "required" reading to "suggested"

1. [SkLearn Estimator Docs](https://scikit-learn.org/stable/developers/develop.html)

>    1. cuML Estimator design follows Sklearn very closely. We will only cover portions where our design differs from this document

Reviewer 2: ```suggestion
   1. cuML Estimator design follows Scikit-learn very closely. We will only cover portions where our design differs from this document
```

   2. Pay close attention to these sections as these topics have caused pain points in the past:
      1. [Instantiation](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      2. [Estimated Attributes](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      3. [`get_params` and `set_params`](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)
      4. [cloning](https://scikit-learn.org/stable/developers/develop.html#estimated-attributes)

> 2. [SkLearn Docstring Guide](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation)

Reviewer 2: ```suggestion
2. [Scikit-learn's Docstring Guide](https://scikit-learn.org/stable/developers/contributing.html#guidelines-for-writing-documentation)
```

>    1. We follow the same guidlines for speficying array-like objects, array shapes, dtypes, and default values

Reviewer 2: ```suggestion
   1. We follow the same guidelines for specifying array-like objects, array shapes, dtypes, and default values
```

## Quick Start Guide

> At a high level, all cuML Estimators must:

Reviewer 2: ```suggestion
At a high level, all standard cuML Estimators must:
```
We should add something to avoid confusion when things from `preprocessing` or other parts that don't need to inherit from `Base`. This is consistent with how afterwards it states `For the majority of estimators, the above steps...`

Author: Not sure I follow your suggestion. If something doesn't inherit from `Base`, it wouldnt be considered an estimator. Are you referring to some of the `preprocessing` classes that dont exactly fit the estimator mold?

Author: Discussed with Dante. Resolving

1. Inherit from `cuml.common.base.Base`
   ```python
   from cuml.common.base import Base

   class MyEstimator(Base):
      ...
   ```
2. Follow the Scikit-learn estimator guidelines found [here](https://scikit-learn.org/stable/developers/develop.html)
3. Include the `Base.__init__()` arguments available in the new Estimator's `__init__()` 
   ```python
   class MyEstimator(Base):

      def __init__(self, *, extra_arg=True, handle=None, verbose=False, output_type=None):
         super().__init__(handle=handle, verbose=verbose, output_type=output_type)
         ...
   ```
4. Declare each array-like attribute the new Estimator will compute as a class variable for automatic array type conversion
   ```python
   from cuml.common.array_descriptor import CumlArrayDescriptor

   class MyEstimator(Base):

      labels_ = CumlArrayDescriptor()

      def __init__(self):
         ...
   ```

> 5. Add input and return type annotations to public API functions OR wrap those functions explicitly with conversion decorators (see [Estimator Methods](#estimator-methods) for non-standard use cases)

Reviewer 2: Perhaps a small example here of a non-standard use case?

Author: I was trying to keep the quick start guide short so instead of adding in a non-standard use case, I linked to one later in the document. Do you feel strongly that it should be included here?

Author: Changed it to a more specific section header.

   ```python
   class MyEstimator(Base):
      
      def fit(self, X) -> "MyEstimator":
         ...

      def predict(self, X) -> CumlArray:
         ...
   ```
6. Implement `get_param_names()` including values returned by `super().get_param_names()`
   ```python
   def get_param_names(self):
      return super().get_param_names() + [
         "eps",
         "min_samples",
      ]
   ```

> For the majority of estimators, the above steps will be sufficient to correctly work with the cuML library and ensure a consistent API. However, situations may arrise where an estimator differs from the standard pattern and some of the functionality needs to be customized. The remainder of this guide takes a deep dive into the estimator functionality to assist developers when building estimators.

Reviewer 2: ```suggestion
For the majority of estimators, the above steps will be sufficient to correctly work with the cuML library and ensure a consistent API. However, situations may arise where an estimator differs from the standard pattern and some of the functionality needs to be customized. The remainder of this guide takes a deep dive into the estimator functionality to assist developers when building estimators.
```

## Background

Some background is necessary to understand the design of estimators and how to work around any non-standard situations.

### Input and Output Types in cuML

> In cuML we support both ingesting and generating a variety of different n-dimensional array types. Estimators should be able to accept and return any array type. The types that are supported as of release 0.17:

Reviewer 2: ```suggestion
In cuML we support both ingesting and generating a variety of different object types. Estimators should be able to accept and return any array type. The types that are supported as of release 0.17:
```
Dataframes like cuDF and Pandas are not n-dimensional array types

 - cuDF DataFrame or Series
 - Pandas DataFrame or Series
 - NumPy Arrays
 - Numba Device Arrays
 - CuPy arrays
 - CumlArray type (Internal to the `cuml` API only.)

> When converting between types, it's important to minimize the CPU<->GPU type conversions as much as possible. Conversions such as NumPy -> CuPy or Numba -> Pandas DataFrame will incur a performance penalty as memory is copied from device to host or viceversa.

Reviewer 2: ```suggestion
When converting between types, it's important to minimize the CPU<->GPU type conversions as much as possible. Conversions such as NumPy -> CuPy or Numba -> Pandas DataFrame will incur a performance penalty as memory is copied from device to host or vice-versa.
```

> Converting between types of the same device, i.e. CPU<->CPU or GPU<->, do not have as significant of a penalty, though they may still increase memory usage.

Author: This is something I am not aware of. Are you saying the Numpy -> cuDF Dataframe is one of the worst conversions in terms of performance?

> Converting between types of the same device, i.e. CPU<->CPU or GPU<->, do not have as significant of a penalty, though they may still increase memory usage.

Author: Got it. Resolving

Finally, conversions between Numba<->CuPy<->CumlArray incur the least amount of overhead since only the device pointer is moved from one class to another.

Internally, all arrays should be converted to `CumlArray` as much as possible since it is compatible with all output types and can be easily converted.

> ### Speficying the Array Output Type

Reviewer 2: ```suggestion
### Specifying the Array Output Type
```

Users can choose which array type should be returned by cuml by either: 
1. Individually setting the output_type property on an estimator class (i.e `Base(output_type="numpy")`)
2. Globally setting the `cuml.global_output_type`

> 3. Temporarily setting the `cuml.global_output_type` via the `cuml.using_output_type` context manager

Reviewer 2: I don't think it is needed to state here that the context manager sets the global output_type, that's an internal implementation detail. The context manager works like a typical python context manager.

Author: Are you suggesting I should remove `#3`?

Author: Ok, resolving.

> **Note:** Setting the global output type <u>will take precedence over any value in `Base.output_type`</u>

Reviewer 2: Is this true for both the context manager or by setting it with `cuml.global_output_type`? Not clear from the wording

Author: Updated to be more clear

Changing the array output type will alter the return value of estimator functions (i.e. `predict()`, `transform()`), and the return value for array-like estimator attributes (i.e. `my_estimator.classes_` or `my_estimator.coef_`)

> All output_types (including `cuml.global_output_type`) are specfied using an all lowercase string. These strings can be passed in an estimators constructor or via `cuml.set_global_output_type` and `cuml.using_output_type`. Accepted values are:

Reviewer 2: ```suggestion
All output_types (including `cuml.global_output_type`) are specified using an all lowercase string. These strings can be passed in an estimators constructor or via `cuml.set_global_output_type` and `cuml.using_output_type`. Accepted values are:
```

 - `None`: (Default) No global value set. Will use individual values from estimators output_type
 - `"input"`: Similar to `None`. Will mirror the same type as any array passed into the estimator

>  - `"mirror"`: Special value for handling internal `cuml` API calls. Will mirror the input value set directly

Reviewer 2: what is the difference in behavior between these 2? If `mirror` is only for internal API calls, I would put it in the end separated, stating that it is for developer use only while the rest are user accessible

Author: Moved mirror to a note at the end to indicate how special it is

   - Note: Cannot be directly set by users
 - `"numba"`: Returns Numba Device Arrays
 - `"numpy"`: Returns Numpy Arrays
 - `"cudf"`: Returns cuDF DataFrame if cols > 1, else cuDF Series
 - `"cupy"`: Returns CuPy Device Arrays

### Ingesting Arrays

When the input array type isn't know, the correct and safest way to ingest arrays is using `cuml.common.input_to_cuml_array`. This method can handle all supported types, is capable of checking the array order, can enforce a specific dtype, and can raise errors on incorrect array sizes:

```python
def fit(self, X):
    cuml_array, dtype, cols, rows = input_to_cuml_array(X, order="K")
    ...
```

### Returning Arrays

> The `CumlArray` class can convert to any supported array type using the `to_output(output_type: str)` method. However, doing this explicitly is almost never needed in practice and **should be avoided** in favor of automatic conversions described below.

Author: Done

The `CumlArrayDescriptor` behaves different when accessed internally (from within one of `cuml`'s functions) vs. externally (for user code outside the cuml module). Internally, it behaves exactly like a normal attribute and will return the previous value set. Externally, the array will get converted to the user's desired output type lazily and repeated conversion will be cached.

#### Defining Array-Like Attributes

To use the `CumlArrayDescriptor` in an estimator, any array-like attributes need to be specified by creating a `CumlArrayDescriptor` as a class variable. 

```python
from cuml.common.array_descriptor import CumlArrayDescriptor

class TestEstimator(cuml.Base):

   # Class variables outside of any function
   my_cuml_array_ = CumlArrayDescriptor()

   def __init__(self, ...):
      ...
```

> This gives the developer full control over which attributes are arrays and the name for the array-like attribute. Array-like attribute **no longer need to follow specific naming conventions** and developers can use any name, assuming it follows the SkLearn attribute naming guidelines and python standards for public/private attributes.

Author: Agreed. I cleaned this up to be more succinct and removed the reference to private/public naming and sklearn naming conventions. That should already be clear.

#### Working with `CumlArrayDescriptor`

Once an `CumlArrayDescriptor` attribute has been defined, developers can treat the attribute as they normally would. Consider the following example estimator:

```python
import cupy as cp
import cuml
from cuml.common.array_descriptor import CumlArrayDescriptor

class TestEstimator(cuml.Base):

   # Class variables outside of any function
   my_cuml_array_ = CumlArrayDescriptor()
   my_cupy_array_ = CumlArrayDescriptor()
   my_other_array_ = CumlArrayDescriptor()

   def __init__(self, ...):

      # Initialize to None (not mandatory)
      self.my_cuml_array_ = None

      # Init with a cupy array
      self.my_cupy_array_ = cp.zeros((10, 10))

   def fit(self, X):
      # Stores the type of `X` and sets the output type if self.output_type == "input"
      self._set_output_type(X)

      # Set my_cuml_array_ with a CumlArray

>       self.my_cuml_array_, _, _, _ = input_to_cuml_array(X, order="K")

Author: You can also do `self.my_cuml_array = input_to_cuml_array(X, order="K").array`

      # Access `my_cuml_array_` normally and set to another attribute
      # The internal type of my_other_array_ will be the result of

>       # CumlArray + CuPy Array = Cupy Array

Author: The more I think about this, the more I dont want to get into this topic at all feel like it should be avoided with more obvious types. Thoughts?

>       # CumlArray + CuPy Array = Cupy Array

Author: I will remove it from here. Potentially a new document on CumlArray and how it can be used with cupy would be helpful.

      self.my_other_array_ = self.my_cuml_array_ + self.my_cupy_array_

      return self
```

Just like any normal attribute, `CumlArrayDescriptor` attributes will return the same value that was set into the attribute _unless accessed externally_ (more on that below). However, developers can convert the type of an array-like attribute by using the `cuml.global_output_type` functionality and reading from the attribute. For example, we could add a `score()` function to `TestEstimator`:

```python
def score(self):

   # Set the global output type to numpy
   with cuml.using_output_type("numpy"):
      # Accessing my_cuml_array_ will return a numpy array and
      # the result can be returned directly
      return np.sum(self.my_cuml_array_, axis=0)
``` 

This has the same benefits of lazy conversion and caching as when descriptors are used externally. 

#### CumlArrayDescriptor External Functionality

> Externally, when users read from a `CumlArrayDescriptor` attribute, the array data will be converted to the correct output type _lazily_ when the attribute is read from. For example, building off the above `TestEstimator`:

Author: I added a section to the intro of 'Estimator Array-Like Attributes'. Let me know what you think.

```python
my_est = TestEstimator()

# Print the default output_type and value for `my_cuml_array_`
# By default, `output_type` is set to `cuml.global_output_type`
# If `cuml.global_output_type == None`, `output_type` is set to "input"
print(my_est.output_type) # Output: "input"
print(my_est.my_cuml_array_) # Output: None
print(my_est.my_other_array_) # Output: AttributeError! my_other_array_ was never set

# Call fit() with a numpy array as the input
np_arr = np.ones((10,))
my_est.fit(np_arr) # This will load data into attributes

# `my_cuml_array_` was set internally as a CumlArray. Externally, we can check the type
print(type(my_est.my_cuml_array_)) # Output: Numpy (saved from the input of `fit`)

# Calling fit again with cupy arrays, will have a similar effect
my_est.fit(cp.ones((10,)))
print(type(my_est.my_cuml_array_)) # Output: CuPy

# Setting the `output_type` will change all descriptor properties
# and ignore the input type
my_est.output_type = "cudf"

# Reading any of the attributes will convert the type lazily

> print(type(my_est.my_cuml_array_)) # Output: cudf

Reviewer 2: ```suggestion
print(type(my_est.my_cuml_array_)) # Output: cuDF object
```

# Setting the global_output_type, overrides the estimator output_type attribute
with cuml.using_output_type("cupy"):
  print(type(my_est.my_cuml_array_)) # Output: cupy

# Once the global_output_type is restored, we return to the estimator output_type

> print(type(my_est.my_cuml_array_)) # Output: Cudf. Using a cached value!

Reviewer 2: ```suggestion
print(type(my_est.my_cuml_array_)) # Output: cuDF. Using a cached value!
```

The full set of return types rules that will be applied by the `Base` metaclass are:

| Return Type | Converts Array Type? | Common Methods | Notes | 
| :---------: | :-----------: | :----------- | :----------- |
| `Base` | No | `fit()` | Any type that `isinstance` of `Base` will work |
| `CumlArray` | Yes | `predict()`, `transform()` | Functions can return any array-like object (`np.ndarray`, `cp.ndarray`, etc. all accepted) |
| `SparseCumlArray` | Yes | `predict()`, `transform()` | Functions can return any sparse array-like object (`scipy`, `cupyx.scipy` sparse arrays accepted) |
| `dict`, `tuple`, `list` or `typing.Union` | Yes |  | Functions must return a generic object that contains an array-like object. No sparse arrays are supported |

Simply setting the return type of a method is all that is necessary to automatically convert the return type (with the added benefit of adding more information to the code). Below are some before/after examples to show simple methods using automatic array conversion.

##### `fit()`

> <!-- <u>Before:</u>

Author: I have added a before/after 0.17 section to the appendix and only show after examples now.

```python

> @with_cupy_rmm

Author: I have added a before/after 0.17 section to the appendix and only show after examples now.

def fit(self, X):
   # Set the base input attributes
   self._set_base_attributes(output_type=X, n_features=X)

   self.coef_ = input_to_cuml_array(X, order="K").array

   return self
```

<u>After:</u>

```python
def fit(self, X) -> "KMeans:
   self.coef_ = input_to_cuml_array(X, order="K").array

   return self
``` -->

<table>
   <thead>
      <tr>
         <th>Before</th>
         <th>After</th>
      </tr>
   </thead>
<tbody>
   <tr>
   <td>

```python
@with_cupy_rmm
def fit(self, X):
   # Set the base input attributes
   self._set_base_attributes(output_type=X, n_features=X)

   self.coef_ = input_to_cuml_array(X, order="K").array

   return self
```

   </td>
   <td stype="text-align: top">

```python
 
def fit(self, X) -> "KMeans":

   self.coef_ = input_to_cuml_array(X, order="K").array

   return self
```

   </td>
   </tr>
</tbody>
</table>

**Notes:**
 - Any type that derives from `Base` can be used as the return type for `fit()`. In python, to indicate returning `self` from a function, class type can be surrounded in quotes to prevent an import error.
 - `@with_cupy_rmm` is no longer needed. This is automatically applied for every public method of estimators

>  - `self._set_base_attributes()` no longer needs to be called. 

Author: I have added a before/after 0.17 section to the appendix and only show after examples now.

##### `predict()`

<!-- ```python
@with_cupy_rmm
def predict(self, X, y):
   # Determine the output type and dtype
   out_type = self._get_output_type(y)

   # Convert to CumlArray
   X_m = input_to_cuml_array(X, order="K").array

   # Do some calculation with cupy
   X_m = cp.asarray(X_m) + cp.ones(X_m.shape)

   # Convert back to CumlArray
   X_m = CumlArray(X_m)

   # Convert the CudaArray to the desired output
   return X_m.to_output(output_type=out_type)
```

**After: `predict()`**
```python
def predict(self, X) -> CumlArray:
   # Convert to CumlArray
   X_m = input_to_cuml_array(X, order="K").array

   # Call a cuda function
   X_m = cp.asarray(X_m) + cp.ones(X_m.shape)

   # Directly return a cupy array
   return X_m
``` -->

<table>
   <thead>
      <tr>

>          <th>Before</th>

Author: I have added a before/after 0.17 section to the appendix and only show after examples now.

   X_m = CumlArray(X_m)

   # Convert the CudaArray to the desired output
   return X_m.to_output(output_type=out_type)
```
   </td>
   <td stype="text-align: top">

```python
 
def predict(self, X) -> CumlArray:

   # Convert to CumlArray
   X_m = input_to_cuml_array(X, order="K").array

   # Call a cuda function
   X_m = cp.asarray(X_m) + cp.ones(X_m.shape)

   # Directly return a cupy array
   return X_m
```

   </td>
   </tr>
</tbody>
</table>

**Notes:**
 - `@with_cupy_rmm` is no longer needed. This is automatically applied for every public method of estimators
 - `self._get_output_type()` no longer needs to be called. The output type is determined automatically
 - Its not necessary to convert to `CumlArray` and casting with `to_output` before returning. This function directly returned a `cp.ndarray` object. Any array-like object can be returned. 

#### Option 2: Manual Estimator Method Decoration

While the automatic converions from type annotations works for many estimator functions, sometimes its necessary to explicitly decorate an estimator method. This allows developers greater flexibility over the input argument, output type and output dtype.

> Which decorator to use for an estimator function is determed by 2 factors:

Reviewer 2: ```suggestion
Which decorator to use for an estimator function is determined by 2 factors:
```

An in-depth discussion of how these decorators work, when each should be used, and their default options can be found in the Appendix. For now, we will show an example method that uses a non-standard input argument name, and also requires converting the array dtype:

##### `predict()`

<!-- **Before:**

```python
@with_cupy_rmm
def predict(self, X_in):
   # Determine the output_type
   out_type = self._get_output_type(X_in)
   out_dtype = self._get_target_dtype()

   # Convert to CumlArray
   X_m = input_to_cuml_array(X_in, order="K").array

   # Call a cuda function
   X_m = cp.asarray(X_m) + cp.ones(X_m.shape)

   # Convert back to CumlArray
   X_m = CumlArray(X_m)

   # Convert the CudaArray to the desired output and dtype
   return X_m.to_output(output_type=out_type, output_dtype=out_dtype)
```

**After:**

```python
@cuml.internals.api_base_return_array(input_arg="X_in", get_output_dtype=True)
def predict(self, X):
   # Convert to CumlArray
   X_m = input_to_cuml_array(X, order="K").array

   # Call a cuda function
   X_m = cp.asarray(X_m) + cp.ones(X_m.shape)

   # Return the cupy array directly
   return X_m
``` -->

<table>
   <thead>
      <tr>

>          <th>Before</th>

Author: Agreed. It was a product of the evolution from a presentation to permanent guide in the code. I have added a before/after 0.17 section to the appendix and only show after examples now.

 - Its not necessary to convert to `CumlArray` and casting with `to_output` before returning. This function directly returned a `cp.ndarray` object. Any array-like object can be returned.
 - Specifying `get_output_dtype=True` in the decorator argument instructs the decorator to also calculate the dtype in addition to the output type.

## Do's And Don'ts

### **Do:** Add Typing Information to Estimator Functions

Adding the return type to estimator functions will allow the `Base` meta-class to automatically decorate functions based on their return type.

**Do this:**
```python
def fit(self, X, y, convert_dtype=True) -> "KNeighborsRegressor":
def predict(self, X, convert_dtype=True) -> CumlArray:
def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity') -> SparseCumlArray:
def predict(self, start=0, end=None, level=None) -> typing.Union[CumlArray, float]:
```

**Not this:**
```python
def fit(self, X, y, convert_dtype=True):
def predict(self, X, convert_dtype=True):
def kneighbors_graph(self, X=None, n_neighbors=None, mode='connectivity'):
def predict(self, start=0, end=None, level=None):
```

> ### **Do:** Return Array-Like Objects Directly

Author: I have added a before/after 0.17 section to the appendix and only show after examples now.

There is no need to convert the array type before returning it. Simply return any array-like object and it will be automatically converted

**Do this:**
```python
def predict(self) -> CumlArray:
   cp_arr = cp.ones((10,))

   return cp_arr
```

**Not this:**
```python
def predict(self, X, y) -> CumlArray:
   cp_arr = cp.ones((10,))

>    # Dont be tempted to use `CumlArray(cp_arr)` here either

Reviewer 2: ```suggestion
   # Don't be tempted to use `CumlArray(cp_arr)` here either
```

      return 1 if self.method_name == "type1" else 0

   def fit(self, X) -> "TestEstimator":

      # Call external code from Cython
      my_external_func(X.ptr, <int>self._method_int())

      return self
```

**Not this:**
```python
class TestEstimator(cuml.Base):

   def __init__(self, method_name: str, ...):
      super().__init__(...)

      self.method_name = 1 if method_name == "type1" else 0

   def fit(self, X) -> "TestEstimator":

      # Call external code from Cython
      my_external_func(X.ptr, <int>self.method_name)

      return self
```

## Additional Information

This section contains more in-depth information about the decorators and descriptors to help developers understand whats going on behind the scenes

### Estimator Array-Like Attributes

#### Automatic Decoration Rules

Adding decorators to every estimator function just to use the decorator default values would be very repetitive and unnecessary. Because most of estimator functions follow a similar pattern, a new meta-class has been created to automatically decorate estimator functions based off their return type. This meta class will decorate functions according to a few rules:

> 1. If a functions has been manually decorated, it will not be automatically decoated

Reviewer 2: ```suggestion
1. If a functions has been manually decorated, it will not be automatically decorated
```

# Output: "cupy"

# The input value can be accessed
print(my_est.__dict__["my_cuml_array_"].get_input_value())
# Output: CumlArray ...
```

### Estimator Methods

#### Common Functionality

All of these decorators perform the same basic steps with a few small differences. The common steps performed by each decorator is:

1. Set `cuml.global_output_type = "mirror"`
   1. When `"mirror"` is used as the global output type, that indicates we are in an internal cuML API call. The `CumlArrayDescriptor` keys off this value to change between internal and external functionality
2. Set CuPy allocator to use RMM
   1. This replaces the existing decorator `@with_cupy_rmm`
   2. Unlike before, the CuPy allocator is only set once per API call
3. Set the estimator input attributes. Can be broken down into 3 steps:
   1. Set `_input_type` attribute
   2. Set `target_dtype` attribute
   3. Set `n_features` attribute
4. Call the desired function
5. Get the estimator output type. Can be broken down into 2 steps:
   1. Get `output_type`
   2. Get `output_dtype`
6. Convert the return value
   1. This will ultimately call `CumlArray.to_output(output_type=output_type, output_dtype=output_dtype)

> While the above list of steps may seem excessive for every call, most functions follow this general form, but may skip a few steps depending on a couple of factors. For example, Step #3 is necessary on functions that modify the estimator's estimated attributes, such as `fit()`, but is not necessary for functions like `predict()` or `transform()`. And Step #5/6 are only necessary when returning array-like objects and are ommitted when returning any other type.

Reviewer 2: ```suggestion
While the above list of steps may seem excessive for every call, most functions follow this general form, but may skip a few steps depending on a couple of factors. For example, Step #3 is necessary on functions that modify the estimator's estimated attributes, such as `fit()`, but is not necessary for functions like `predict()` or `transform()`. And Step #5/6 are only necessary when returning array-like objects and are omitted when returning any other type.
```

Functionally, you can think of these decorators equivalent to the following pseudocode:
```python
def my_func(self, X):
   with cuml.using_ouput_type("mirror"):
      with cupy.cuda.cupy_using_allocator(rmm.rmm_cupy_allocator):
         # Set the input properties
         self._set_base_attributes(output_type=X, n_features=X)

         # Do actual calculation returning an array-like object
         ret_val = self._my_func(X)

         # Get the output type
         output_type = self._get_output_type(X)

         # Convert array-like to CumlArray
         ret_val = input_to_cuml_array(ret_val, order="K").array

         # Convert CumlArray to desired output_type
         return ret_val.to_output(output_type)
```

> Keep the above psuedocode in mind when working with these decorators since their goal is to replace many of these repetitive functions.

Reviewer 2: ```suggestion
Keep the above pseudocode in mind when working with these decorators since their goal is to replace many of these repetitive functions.
```
